contracts/YieldStEthLever.sol

```diff
@@ -66,279 +75,391 @@ interface YieldLadle {
         bool success = fyToken.flashLoan(
             this, // Loan Receiver
             address(fyToken), // Loan Token
-            investAmount, // Loan Amount
-            abi.encode(0, abi.encode(borrowAmount, vaultId))
+            borrowAmount, // Loan Amount
+            data
         );
-        require(success, "Failed to flash loan");
+        if (!success) revert FlashLoanFailure();
         giver.give(vaultId, msg.sender);
-        IERC20(address(fyToken)).transfer(
-            msg.sender,
-            IERC20(address(fyToken)).balanceOf(address(this))
-        );
-        return vaultId;
+        // We put everything that we borrowed into the vault, so there can't be
+        // any FYTokens left. Check:
+        // assert(IERC20(address(fyToken)).balanceOf(address(this)) == 0);
     }
 
+    /// @notice Called by a flash lender, which can be `fyToken` or
+    ///     `flashJoin2` (for Weth). The primary purpose is to check conditions
+    ///     and route to the correct internal function.
+    ///
+    ///     This function reverts if not called through a flashloan initiated
+    ///     by this contract.
+    /// @param initiator The initator of the flash loan, must be `address(this)`.
+    /// @param borrowAmount The amount of fyTokens received.
+    /// @param fee The fee that is subtracted in addition to the borrowed
+    ///     amount when repaying.
+    /// @param data The data we encoded for the functions. Here, we only check
+    ///     the first byte for the router.
     function onFlashLoan(
         address initiator,
-        address token,
-        uint256 amount, // Amount of FYToken received
+        address, // The token, not checked as we check the lender address.
+        uint256 borrowAmount,
         uint256 fee,
-        bytes memory data
+        bytes calldata data
     ) external returns (bytes32) {
-        address thisAdd = address(this);
-        require(
-            initiator == thisAdd,
-            "FlashBorrower: Untrusted loan initiator"
-        );
-
-        (uint128 status, bytes memory data2) = abi.decode(
-            data,
-            (uint128, bytes)
-        );
-        if (status == 0) {
-            leverUp(
-                amount, // Amount of FYToken received
-                fee,
-                data2
-            );
-        } else if (status == 1) {
-            doRepay(amount, fee, data2);
-        } else if (status == 2) {
-            doClose(amount, fee, data2);
+        // Test that the flash loan was sent from the a contract and that this
+        // contract was the initiator.
+        if (
+            (msg.sender != address(fyToken) &&
+                msg.sender != address(flashJoin2)) || initiator != address(this)
+        ) revert FlashLoanFailure();
+
+        // Decode the operation to execute and then call that function.
+        bytes1 status = data[0];
+        if (status == 0x01) {
+            leverUp(borrowAmount, fee, data);
```

Comment:
Do all the decoding in this function, as soon as `data` is received and we've verified the sources.